File: apps/ios/Sources/CrossyUI/RoomAnalysis.swift
// The post-game analysis surface's data, as the render layer reads it (the
// RosterMember pattern, ARCHITECTURE.md AD-2: CrossyUI names its own plain types,
// and the protocol twins stay in their ring). The composition root fetches
// `GET /games/{id}/analysis` through CrossyAPI and maps its `AnalysisView` into
// these before the room ever sees it, so CrossyUI keeps importing only CrossyStore
// and CrossyDesign.
//
// The bundle is first-correct truth (design/post-game/ANALYSIS.md, engine
// `solveTrace`): `owners` is who solved each cell FIRST, the same attribution the
// web mosaic and legend paint, distinct from the live event log's last-writer
// `by`. It carries userIds, cells, and numbers only, never a letter (INV-6): the
// server strips the trace to this shape, and this type has nowhere to hold a
// solution value by construction.

import Foundation
import Observation

/// The analysis bundle for one completed game, render-ready. Times are relative
/// seconds from the solve's start (design/post-game/ANALYSIS.md); `owners` maps a
/// cell index to the userId who first got it right.
public struct RoomAnalysis: Equatable, Sendable {
    /// Cell index to first-correct writer's userId. The mosaic's colors and the
    /// legend's roster both read this.
    public let owners: [Int: String]
    public let momentum: RoomMomentum
    /// The room's longest pause and the burst that broke it, or nil when the trace
    /// is too short to have a gap (fewer than two fills).
    public let turningPoint: RoomTurningPoint?
    /// The solver superlatives, in the wire's ladder-rank order
    /// (design/post-game/TITLES.md): at most one per solver, empty for a solo solve
    /// (the solo rule) or an older API that predates titles.
    public let titles: [RoomTitle]

    public init(
        owners: [Int: String],
        momentum: RoomMomentum,
        turningPoint: RoomTurningPoint?,
        titles: [RoomTitle]
    ) {
        self.owners = owners
        self.momentum = momentum
        self.turningPoint = turningPoint
        self.titles = titles
    }

    /// Distinct solvers who own at least one square (the stat trio's "Solvers").
    public var solverCount: Int { Set(owners.values).count }

    /// Total squares with a first-correct owner (the stat trio's "Squares").
    public var entryCount: Int { owners.count }

    /// The solve span as `M:SS` (the stat trio's "Time"): the momentum duration,
    /// the reach from the first fill to the last (design/post-game/ANALYSIS.md),
    /// which is what the web panel labels Time.
    public var durationLabel: String { CrossyUI.formatMSS(momentum.durationSeconds) }
}

/// A seconds count as `M:SS`, or `H:MM:SS` past an hour, matching the web's
/// formatMSS (apps/web/src/ui/analysisReadout.ts) byte for byte: seconds floored,
/// the seconds and (when hours show) minutes fields zero-padded, negatives and
/// non-finite input clamped to "0:00" so a degenerate span never reads "NaN:NaN".
/// The one CrossyUI moment formatter, so the Analysis header (RoomAnalysis) and the
/// solver-title claims (TitleLadder) render the same span identically, and both
/// agree with the web digit for digit. Pure so it pins headlessly.
func formatMSS(_ seconds: Double) -> String {
    let safe = seconds.isFinite ? seconds : 0
    let total = max(0, Int(safe.rounded(.down)))
    let hours = total / 3600
    let minutes = (total % 3600) / 60
    let secs = total % 60
    let pad = { (n: Int) in String(format: "%02d", n) }
    return hours > 0 ? "\(hours):\(pad(minutes)):\(pad(secs))" : "\(minutes):\(pad(secs))"
}

/// The solving-tempo ribbon's data: a fixed-length, peak-normalized intensity
/// series and the span it covers (engine `momentum`, design/post-game/ANALYSIS.md).
public struct RoomMomentum: Equatable, Sendable {
    /// The solve span in seconds (0 for an empty or instant solve).
    public let durationSeconds: Double
    /// 40 samples in [0, 1], each a bucket's fill count over the busiest bucket's
    /// (engine constant `MOMENTUM_SAMPLES`). All zero when nothing was filled.
    public let samples: [Double]

    public init(durationSeconds: Double, samples: [Double]) {
        self.durationSeconds = durationSeconds
        self.samples = samples
    }

    /// True when any bucket carried a fill: the ribbon has a shape to draw, rather
    /// than the flat quiet line a short solve leaves.
    public var hasSignal: Bool { samples.contains { $0 > 0 } }
}

/// One solver superlative (design/post-game/TITLES.md): the wire's award, carried
/// verbatim. `key` is the lowercase-kebab ladder key ("saboteur"); the display table
/// (TitleLadder) decides what it knows, so an unknown key from a grown ladder is
/// skipped at render, never dropped here (forward compatibility, PROTOCOL.md §12).
public struct RoomTitle: Equatable, Sendable {
    public let userId: String
    /// The pinned lowercase-kebab title key, verbatim from the wire.
    public let key: String
    /// The rung's own count (overwrites, whole seconds, squares), or nil for a rung
    /// that cites none. Never a letter (INV-6).
    public let evidence: Int?

    public init(userId: String, key: String, evidence: Int?) {
        self.userId = userId
        self.key = key
        self.evidence = evidence
    }
}

/// The room's longest pause and the burst that ended it (engine `TurningPoint`):
/// the ribbon shades the stall span and marks where solving picked back up.
public struct RoomTurningPoint: Equatable, Sendable {
    /// The largest gap between consecutive fills, in seconds.
    public let stallSeconds: Double
    /// The relative time, in seconds, of the fill that ended the gap.
    public let breakSeconds: Double
    /// Fills within the 30-second window after the break (engine `BURST_WINDOW_MS`).
    public let burst: Int

    public init(stallSeconds: Double, breakSeconds: Double, burst: Int) {
        self.stallSeconds = stallSeconds
        self.breakSeconds = breakSeconds
        self.burst = burst
    }
}

/// The analysis fetch's state machine, the thin observable the solve screen drives.
/// The bundle is fetched exactly once per completed room, off an injected async
/// closure (the composition root closes over the REST client and game id, keeping
/// CrossyUI out of the REST ring). A completion can be observed over the socket a
/// beat before the session has flushed `completed_at` to Postgres, so the endpoint
/// 404s for a short window right after a live finish; the load retries a few times
/// before it calls the game `absent` (the web client's completion-race guard,
/// apps/web completionAttribution).
@available(iOS 17.0, macOS 14.0, *)
@MainActor
@Observable
public final class AnalysisModel {
    /// The fetch's four states: not yet asked, in flight, resolved, or given up
    /// (a 404 past the retries, or a genuine absence).
    public enum Phase: Equatable, Sendable {
        case idle
        case loading
        case ready(RoomAnalysis)
        case absent
    }

    public private(set) var phase: Phase = .idle

    @ObservationIgnored private var task: Task<Void, Never>?

    public init() {}

    /// The resolved bundle, or nil until it lands.
    public var bundle: RoomAnalysis? {
        if case let .ready(bundle) = phase { return bundle }
        return nil
    }

    /// Kick the one fetch for this room. Idempotent: a second call while a fetch is
    /// in flight or already resolved is a no-op, so both the completion edge and a
    /// tab-open can ask without racing. `fetch` returns nil on any failure (a 404,
    /// transport weather, a decode fault); the retries cover the completion race,
    /// after which `absent` stands.
    public func load(
        tries: Int = 3,
        delay: Duration = .milliseconds(700),
        _ fetch: @escaping () async -> RoomAnalysis?
    ) {
        guard case .idle = phase else { return }
        phase = .loading
        task?.cancel()
        task = Task { @MainActor [weak self] in
            for attempt in 0..<max(1, tries) {
                if Task.isCancelled { return }
                if let bundle = await fetch() {
                    guard !Task.isCancelled else { return }
                    self?.phase = .ready(bundle)
                    return
                }
                if attempt < tries - 1 {
                    try? await Task.sleep(for: delay)
                }
            }
            guard !Task.isCancelled else { return }
            self?.phase = .absent
        }
    }
}

File: apps/android/ui/src/main/kotlin/crossy/ui/RoomAnalysis.kt
// The post-game analysis surface's data, as the render layer reads it (twin of apps/ios
// RoomAnalysis.swift; the RosterMember pattern, AAD-2: :ui names its own plain types and the
// protocol twins stay in their ring). The composition root fetches GET /games/{id}/analysis
// through :api and maps its AnalysisView into these before the room ever sees it.
//
// The bundle is first-correct truth (design/post-game/ANALYSIS.md, engine solveTrace): `owners`
// is who solved each cell FIRST, the same attribution the web mosaic and legend paint, distinct
// from the live event log's last-writer `by`. It carries userIds, cells, and numbers only, never
// a letter (INV-6): the server strips the trace to this shape, and this type has nowhere to hold
// a solution value by construction.

package crossy.ui

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.delay

/** The analysis bundle for one completed game, render-ready. Times are relative seconds from the
 *  solve's start; `owners` maps a cell index to the userId who first got it right. */
data class RoomAnalysis(
    val owners: Map<Int, String>,
    val momentum: RoomMomentum,
    /** The room's longest pause and the burst that broke it, or null when the trace is too short
     *  to have a gap (fewer than two fills). */
    val turningPoint: RoomTurningPoint?,
    /** The solver superlatives, in the wire's ladder-rank order (design/post-game/TITLES.md): at most
     *  one per solver, empty for a solo solve (the solo rule) or an older API that predates titles. */
    val titles: List<RoomTitle>,
    /** The sittings partition (design/post-game/SITTINGS.md, D29), or null from an older cached
     *  bundle, which renders exactly as today: no context suffix, no seam ticks. */
    val sittings: RoomSittings? = null,
) {
    /** Distinct solvers who own at least one square (the stat trio's "Solvers"). */
    val solverCount: Int get() = owners.values.toSet().size

    /** Total squares with a first-correct owner (the stat trio's "Squares"). */
    val entryCount: Int get() = owners.size

    /** The solve span as `M:SS` (the stat trio's "Time"): the momentum duration, the reach from
     *  the first fill to the last, which is what the web panel labels Time. Active seconds once the
     *  server re-bases the bundle (D29); the label logic does not change. */
    val durationLabel: String get() = formatMSS(momentum.durationSeconds)

    /** The sitting-count context beside the headline Time ("24:13 · 2 sittings", owner ruling, D29):
     *  context, never a second stat, so it renders only at two or more sittings. A single-sitting
     *  game, an older bundle, and a degenerate count all read exactly as today: null, no suffix. */
    val sittingCountSuffix: String?
        get() {
            val count = sittings?.count ?: return null
            return if (count >= 2) "$count sittings" else null
        }

    companion object {
        /** A seconds count as `M:SS`, or `H:MM:SS` past an hour, matching the web's formatMSS
         *  (apps/web/src/ui/analysisReadout.ts) digit for digit: seconds floored, the seconds and
         *  (when hours show) minutes fields zero-padded, negatives and non-finite input clamped to
         *  "0:00" so a degenerate span never reads "NaN". Pure so it pins headlessly. */
        fun formatMSS(seconds: Double): String {
            val safe = if (seconds.isFinite()) seconds else 0.0
            val total = maxOf(0, kotlin.math.floor(safe).toInt())
            val hours = total / 3600
            val minutes = (total % 3600) / 60
            val secs = total % 60
            return if (hours > 0) {
                "$hours:${"%02d".format(minutes)}:${"%02d".format(secs)}"
            } else {
                "$minutes:${"%02d".format(secs)}"
            }
        }
    }
}

/** The solving-tempo ribbon's data: a fixed-length, peak-normalized intensity series and the span
 *  it covers (engine `momentum`). */
data class RoomMomentum(
    /** The solve span in seconds (0 for an empty or instant solve). */
    val durationSeconds: Double,
    /** 40 samples in [0, 1], each a bucket's fill count over the busiest bucket's (engine
     *  MOMENTUM_SAMPLES). All zero when nothing was filled. */
    val samples: List<Double>,
) {
    /** True when any bucket carried a fill: the ribbon has a shape to draw, rather than the flat
     *  quiet line a short solve leaves. */
    val hasSignal: Boolean get() = samples.any { it > 0 }
}

/** One solver superlative (design/post-game/TITLES.md): the wire's award, carried verbatim. `key` is
 *  the lowercase-kebab ladder key ("saboteur"); the display table (TitleLadder) decides what it knows,
 *  so an unknown key from a grown ladder is skipped at render, never dropped here (forward
 *  compatibility, PROTOCOL.md §12). Twin of the iOS RoomTitle. */
data class RoomTitle(
    val userId: String,
    /** The pinned lowercase-kebab title key, verbatim from the wire. */
    val key: String,
    /** The rung's own count (overwrites, whole seconds, squares), or null for a rung that cites
     *  none. Never a letter (INV-6). */
    val evidence: Int?,
)

/** The sittings partition, render-ready (design/post-game/SITTINGS.md, D29): the count for the
 *  headline's context suffix, the spans for the ribbon's seam ticks, and the wall span for flavor
 *  copy. All seconds ride the bundle's active axis, the same axis as `RoomMomentum.durationSeconds`,
 *  so a seam places by lookup. Twin of the iOS RoomSittings. */
data class RoomSittings(
    val count: Int,
    val spans: List<Span>,
    /** The wall-clock trace span (the pre-D29 duration), flavor only, never a competing stat
     *  (owner ruling, D29). */
    val wallSeconds: Double,
) {
    /** One sitting's reach on the active axis, contiguous with its neighbors. */
    data class Span(val startSeconds: Double, val endSeconds: Double)

    /** The interior sitting boundaries on the active axis: `spans[k].endSeconds` for every k but the
     *  last, where the ribbon draws its seam ticks. Empty for a single sitting, so the ribbon renders
     *  exactly as today. */
    val interiorBoundarySeconds: List<Double>
        get() = spans.dropLast(1).map { it.endSeconds }
}

/** The room's longest pause and the burst that ended it (engine `TurningPoint`): the ribbon
 *  shades the stall span and marks where solving picked back up. */
data class RoomTurningPoint(
    /** The largest gap between consecutive fills, in seconds. */
    val stallSeconds: Double,
    /** The relative time, in seconds, of the fill that ended the gap. */
    val breakSeconds: Double,
    /** Fills within the 30-second window after the break (engine BURST_WINDOW_MS). */
    val burst: Int,
)

/** The analysis fetch's state machine, the thin observable the solve screen drives (twin of iOS
 *  AnalysisModel). The bundle is fetched exactly once per completed room, off an injected suspend
 *  closure (the composition root closes over the REST client and game id, keeping :ui out of the
 *  REST ring). A completion can be observed over the socket a beat before the session has flushed
 *  completed_at to Postgres, so the endpoint 404s for a short window right after a live finish;
 *  the load retries a few times before it calls the game absent (the web client's
 *  completion-race guard). */
class AnalysisModel {
    /** The fetch's four states: not yet asked, in flight, resolved, or given up (a 404 past the
     *  retries, or a genuine absence). */
    sealed interface Phase {
        data object Idle : Phase
        data object Loading : Phase
        data class Ready(val bundle: RoomAnalysis) : Phase
        data object Absent : Phase
    }

    var phase: Phase by mutableStateOf(Phase.Idle)
        private set

    /** The resolved bundle, or null until it lands. */
    val bundle: RoomAnalysis? get() = (phase as? Phase.Ready)?.bundle

    /** Kick the one fetch for this room. Idempotent: a second call while a fetch is in flight or
     *  already resolved is a no-op, so both the completion edge and a tab-open can ask without
     *  racing. `fetch` returns null on any failure (a 404, transport weather, a decode fault);
     *  the retries cover the completion race, after which absent stands. Suspends on the caller's
     *  scope (the room's), so leaving the room cancels the walk. */
    suspend fun load(
        tries: Int = 3,
        delayMillis: Long = 700,
        fetch: suspend () -> RoomAnalysis?,
    ) {
        if (phase != Phase.Idle) return
        phase = Phase.Loading
        repeat(maxOf(1, tries)) { attempt ->
            val bundle = fetch()
            if (bundle != null) {
                phase = Phase.Ready(bundle)
                return
            }
            if (attempt < tries - 1) delay(delayMillis)
        }
        phase = Phase.Absent
    }
}
